File: greathost.js
const EMAIL = process.env.GREATHOST_EMAIL || '';
const PASSWORD = process.env.GREATHOST_PASSWORD || '';
const CHAT_ID = process.env.CHAT_ID || '';
const BOT_TOKEN = process.env.BOT_TOKEN || '';
// === SOCKS5 代理配置 ===
const PROXY_URL = (process.env.PROXY_URL || "").trim();

// 🛑 核心修改：使用 firefox 避开 Chromium 的 SOCKS5 认证限制
const { firefox } = require("playwright");
const https = require('https');

async function sendTelegramMessage(message) {
    return new Promise((resolve) => {
        const url = `https://api.telegram.org/bot${BOT_TOKEN}/sendMessage`;
        const data = JSON.stringify({ chat_id: CHAT_ID, text: message, parse_mode: 'HTML' });
        const options = { 
            method: 'POST', 
            headers: { 
                'Content-Type': 'application/json', 
                'Content-Length': Buffer.byteLength(data) 
            } 
        };
        const req = https.request(url, options, (res) => {
            res.on('data', () => {});
            res.on('end', () => resolve());
        });
        req.on('error', () => resolve());
        req.write(data);
        req.end();
    });
}

(async () => {
    // === 变量定义 ===
    const GREATHOST_URL = "https://greathost.es";    
    const LOGIN_URL = `${GREATHOST_URL}/login`;
    const HOME_URL = `${GREATHOST_URL}/dashboard`;
    const BILLING_URL = `${GREATHOST_URL}/billing/free-servers`;
    
    let proxyStatusTag = "🌐 直连模式";
    let serverStarted = false;

    // --- 1. 严格解析 PROXY_URL ---
    let proxyData = null;
    if (PROXY_URL && PROXY_URL.trim().length > 0) {
        try {
            // 确保有协议头
            let cleanUrl = PROXY_URL.trim();
            if (!cleanUrl.startsWith('socks')) {
                cleanUrl = `socks5://${cleanUrl}`;
            }
            proxyData = new URL(cleanUrl);
            proxyStatusTag = `🔒 代理模式 (${proxyData.host})`;
        } catch (e) {
            console.error("❌ PROXY_URL 解析失败:", e.message);
            proxyStatusTag = "⚠️ 代理配置错误 (退回直连)";
        }
    }

    let browser;
    try {
        console.log(`🚀 任务启动 | 引擎: Firefox | ${proxyStatusTag}`);
        
        // --- 2. 启动 Firefox (这里必须传 proxy 才能生效) ---
        const launchOptions = { headless: true };
        if (proxyData) {
            // 只传服务器地址，避免 Playwright 报认证不支持的错
            launchOptions.proxy = { server: `socks5://${proxyData.host}` };
        }
        browser = await firefox.launch(launchOptions);

        // --- 3. 创建上下文 (全局唯一声明) ---
        const context = await browser.newContext({
            userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0',
            viewport: { width: 1280, height: 720 },
            locale: 'es-ES'
        });

        // --- 4. 注入 SOCKS5 账号密码 (解决认证问题) ---
        if (proxyData && proxyData.username) {
            await context.setHttpCredentials({
                username: proxyData.username,
                password: proxyData.password
            });
            console.log(`🔑 代理认证已注入: ${proxyData.username}`);
        }

        const page = await context.newPage();

        // --- 5. 抹除特征 ---
        await page.addInitScript(() => {
            Object.defineProperty(navigator, 'webdriver', { get: () => false });
        });

        // --- 6. 代理 IP 熔断检查 ---
        if (proxyData) {
            console.log("🌍 [Check] 正在检测代理出口 IP...");
            try {
                // 增加超时，防止代理过慢导致崩溃
                await page.goto("https://api.ipify.org?format=json", { timeout: 60000 });
                const ipInfo = JSON.parse(await page.innerText('body'));
                console.log(`✅ 当前出口 IP: ${ipInfo.ip}`);
            } catch (e) {
                console.error(`❌ 代理检查失败: ${e.message}`);
                throw new Error(`Proxy Check Failed: ${e.message}`);
            }
        }

        // 6. 登录流程
        console.log("🔑 步骤 1: 访问登录页面...");
        await page.goto(LOGIN_URL, { waitUntil: "domcontentloaded" });
        await page.fill('input[name="email"]', EMAIL);
        await page.fill('input[name="password"]', PASSWORD);
        
        console.log("🔑 步骤 2: 提交登录信息...");
        await Promise.all([
            page.click('button[type="submit"]'),
            page.waitForNavigation({ waitUntil: "networkidle" }),
        ]);
        
        if (page.url().includes('login')) {
            throw new Error("登录失败，页面仍留在登录页，请检查凭据或验证码。");
        }
        console.log("✅ 登录成功！");

        // 7. 自动检查服务器并尝试开机
        console.log("📊 步骤 3: 检查服务器运行状态...");
        await page.goto(HOME_URL, { waitUntil: "networkidle" });
        
        // 定位离线标识（根据 GreatHost 实际 DOM 调整）
        const offlineIndicator = page.locator('span.badge-danger:has-text("Offline"), .status-offline').first();
        if (await offlineIndicator.isVisible()) {
            console.log("⚠️ 检测到服务器离线，尝试发送启动指令...");
            const startBtn = page.locator('button:has-text("Start"), .btn-start').first();
            if (await startBtn.isVisible()) {
                await startBtn.click();
                serverStarted = true;
                await page.waitForTimeout(5000); // 等待启动反馈
                console.log("✅ 启动指令已发出");
            }
        } else {
            console.log("🟢 服务器已在运行中。");
        }

        // 8. 续期业务逻辑
        console.log("🔍 步骤 4: 进入 Billing 免费服务器列表...");
        await page.goto(BILLING_URL, { waitUntil: "networkidle" });

        // 点击 "View Details"
        console.log("🔍 步骤 5: 点击 View Details 进入详情页...");
        const detailLink = page.getByRole('link', { name: 'View Details' }).first();
        if (!(await detailLink.isVisible())) {
            throw new Error("未找到 View Details 链接，可能没有有效的免费服务器。");
        }
        await detailLink.click();
        await page.waitForNavigation({ waitUntil: "networkidle" });
        
        const serverId = page.url().split('/').pop() || 'unknown';
        const timeSelector = '#accumulated-time';

        // 获取续期前时长
        const beforeHoursText = await page.textContent(timeSelector).catch(() => "0h");
        const beforeHours = parseInt(beforeHoursText.replace(/[^0-9]/g, '')) || 0;
        console.log(`⏰ 续期前累计时长: ${beforeHours}h`);

        // 检查续期按钮状态
        const renewBtn = page.locator('#renew-free-server-btn');
        const btnContent = await renewBtn.innerHTML();

        const getReport = (icon, title, hours, detail) => {
            return `${icon} <b>GreatHost ${title}</b>\n\n` +
                   `🆔 <b>服务器ID:</b> <code>${serverId}</code>\n` +
                   `⏰ <b>当前时长:</b> ${hours}h\n` +
                   `🚀 <b>开机状态:</b> ${serverStarted ? '✅ 已触发开机' : '运行中'}\n` +
                   `🌐 <b>连接模式:</b> ${proxyStatusTag}\n` + 
                   `📅 <b>时间:</b> ${new Date().toLocaleString('zh-CN', { timeZone: 'Asia/Shanghai' })}\n` +
                   `💡 <b>详情:</b> ${detail}`;
        };

        if (btnContent.includes('Wait')) {
            const waitMatch = btnContent.match(/\d+/);
            const waitTime = waitMatch ? waitMatch[0] : "未知";
            console.log(`⏳ 按钮锁定中，需等待 ${waitTime} 分钟`);
            await sendTelegramMessage(getReport('⏳', '续期冷却中', beforeHours, `还需等待 ${waitTime} 分钟`));
            return;
        }

        // 9. 执行模拟真人点击续期
        console.log("⚡ 步骤 6: 模拟真人点击续期按钮...");
        await page.mouse.wheel(0, 300); // 向下滚动
        await page.waitForTimeout(2000);
        await renewBtn.click({ force: true, delay: 150 });

        // 10. 等待同步并最终校验
        console.log("⏳ 步骤 7: 等待 20 秒处理数据写入...");
        await page.waitForTimeout(20000); 
        await page.reload({ waitUntil: "networkidle" });
        
        const afterHoursText = await page.textContent(timeSelector).catch(() => "0h");
        const afterHours = parseInt(afterHoursText.replace(/[^0-9]/g, '')) || 0;
        console.log(`⏰ 续期后累计时长: ${afterHours}h`);

        // 11. 发送最终通知
        if (afterHours > beforeHours) {
            console.log("🎉 续期成功！");
            await sendTelegramMessage(getReport('🎉', '续期成功', afterHours, `时长已从 ${beforeHours}h 成功增加`));
        } else {
            console.log("✅ 时长未变，可能已达上限或点击未生效。");
            await sendTelegramMessage(getReport('✅', '已检查', afterHours, '目前时长充足或点击受限，建议手动核实'));
        }

    } catch (err) {
        console.error("❌ 脚本运行崩溃:", err.message);
        // 如果不是主动触发的代理熔断，则发送崩溃通知
        if (!err.message.includes("Proxy Check Failed")) {
            await sendTelegramMessage(`🚨 <b>GreatHost 脚本崩溃</b>\n错误原因: <code>${err.message}</code>\n状态: ${proxyStatusTag}`);
        }
    } finally {
        if (browser) {
            console.log("🧹 [Exit] 正在关闭浏览器...");
            await browser.close();
        }
    }
})();

File: greathost.py
import os, re, time, random, requests, json
from datetime import datetime, timezone
from urllib.parse import urlparse
from zoneinfo import ZoneInfo

from seleniumwire import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

# ================= 配置区 =================
EMAIL = os.getenv("GREATHOST_EMAIL", "")
PASSWORD = os.getenv("GREATHOST_PASSWORD", "")
TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "")
TELEGRAM_CHAT_ID = os.getenv("TELEGRAM_CHAT_ID", "")
PROXY_URL = os.getenv("PROXY_URL", "")

# 初始配置名，可以为空
TARGET_NAME_CONFIG = os.getenv("TARGET_NAME", "loveMC") 

STATUS_MAP = {
    "running": ["🟢", "Running"],
    "starting": ["🟡", "Starting"],
    "stopped": ["🔴", "Stopped"],
    "offline": ["⚪", "Offline"],
    "suspended": ["🚫", "Suspended"]
}

# ================= 工具函数 =================
def now_shanghai():
    return datetime.now(ZoneInfo("Asia/Shanghai")).strftime('%Y/%m/%d %H:%M:%S')

def calculate_hours(date_str):
    try:
        if not date_str: 
            return 0
        
        clean_date = re.sub(r'\.\d+Z$', 'Z', str(date_str))
        expiry = datetime.fromisoformat(clean_date.replace('Z', '+00:00'))
        now = datetime.now(timezone.utc)
        
        # 4. 计算小时差
        diff = (expiry - now).total_seconds() / 3600
        
        # 5. 如果差值小于 0，说明已过期，返回 0；否则返回整数小时
        result = max(0, int(diff))
        print(f"🕒 时间计算调试: 原始={date_str} -> 解析后={clean_date} -> 剩余={result}h")
        return result
    except Exception as e:
        print(f"⚠️ 时间解析失败 ({date_str}): {e}")
        return 0

def fetch_api(driver, url, method="GET"):
    script = f"return fetch('{url}', {{method:'{method}'}}).then(r=>r.json()).catch(e=>({{success:false,message:e.toString()}}))"
    res = driver.execute_script(script)
    print(f"📡 API 调用 [{method}] {url}")
    return res

def send_notice(kind, fields):
    titles = {
        "renew_success": "🎉 <b>GreatHost 续期成功</b>",
        "maxed_out": "🈵 <b>GreatHost 已达上限</b>",
        "cooldown": "⏳ <b>GreatHost 还在冷却中</b>",
        "renew_failed": "⚠️ <b>GreatHost 续期未生效</b>",
        "error": "🚨 <b>GreatHost 脚本报错</b>"
    }
    title = titles.get(kind, "‼️ <b>GreatHost 通知</b>")
    body = "\n".join([f"{e} {l}: {v}" for e, l, v in fields])
    msg = f"{title}\n\n{body}\n📅 时间: {now_shanghai()}"
    if TELEGRAM_BOT_TOKEN and TELEGRAM_CHAT_ID:
        try: requests.post(f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage", 
                          data={"chat_id": TELEGRAM_CHAT_ID, "text": msg, "parse_mode": "HTML"}, timeout=5)
        except: pass

# ================= 主流程 =================
def run_task():
    driver = None
    server_id = "未知"
    current_server_name = "未知" # 统一使用此变量名
    login_ip = "Unknown"
    
    try:
        opts = Options()
        opts.add_argument("--headless=new")
        driver = webdriver.Chrome(options=opts, seleniumwire_options={'proxy': {'http': PROXY_URL, 'https': PROXY_URL}} if PROXY_URL else None)
        wait = WebDriverWait(driver, 25)

        # 0. 登入 IP 打印
        try:
            driver.get("https://api.ipify.org?format=json")
            login_ip = json.loads(driver.find_element(By.TAG_NAME, "body").text).get('ip', 'Unknown')
            print(f"🌐 登入 IP: {login_ip}")
        except: pass

        # 1. 登录
        driver.get("https://greathost.es/login")
        wait.until(EC.presence_of_element_located((By.NAME,"email"))).send_keys(EMAIL)
        driver.find_element(By.NAME,"password").send_keys(PASSWORD)
        driver.find_element(By.CSS_SELECTOR,"button[type='submit']").click()
        wait.until(EC.url_contains("/dashboard"))

        # 2. 智能锁定服务器逻辑
        res = fetch_api(driver, "/api/servers")
        server_list = res.get('servers', [])
        
        if not server_list: raise Exception("账号下没有找到任何服务器")

        if TARGET_NAME_CONFIG:
            # 精准匹配
            target_server = next((s for s in server_list if s.get('name') == TARGET_NAME_CONFIG), None)
            if not target_server: raise Exception(f"未找到名称为 '{TARGET_NAME_CONFIG}' 的服务器")
        else:
            # 自动判定
            if len(server_list) == 1:
                target_server = server_list[0]
            else:
                raise Exception(f"账号下存在 {len(server_list)} 个服务器，必须指定 TARGET_NAME")

        server_id = target_server.get('id')
        current_server_name = target_server.get('name') # 获取真实名字
        print(f"✅ 已锁定服务器: {current_server_name}")
        
        # 3. 获取状态
        info = fetch_api(driver, f"/api/servers/{server_id}/information")
        real_status = info.get('status', 'unknown').lower()
        icon, status_name = STATUS_MAP.get(real_status, ["❓", real_status])
        status_disp = f"{icon} {status_name}"

        # 4. 合同预检
        driver.get(f"https://greathost.es/contracts/{server_id}")
        time.sleep(2)
        contract = fetch_api(driver, f"/api/servers/{server_id}/contract")
        before_h = calculate_hours(contract.get('renewalInfo', {}).get('nextRenewalDate'))
        
        btn = wait.until(EC.presence_of_element_located((By.ID, "renew-free-server-btn")))
        if "Wait" in btn.text:
            m = re.search(r"Wait\s+(\d+\s+\w+)", btn.text)
            send_notice("cooldown", [
                ("🖥️", "服务器名称", current_server_name),
                ("⏳", "冷却时间", m.group(1) if m else btn.text),
                ("📊", "当前累计", f"{before_h}h")
            ])
            return

        # 5. 执行续期
        renew_res = fetch_api(driver, f"/api/renewal/contracts/{server_id}/renew-free", method="POST")
        after_h = calculate_hours(renew_res.get('details', {}).get('nextRenewalDate')) or before_h

        # 6. 发送通知 (统一使用 current_server_name)
        if renew_res.get('success') and after_h > before_h:
            send_notice("renew_success", [
                ("🖥️", "服务器名称", current_server_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏰", "增加时间", f"{before_h} ➔ {after_h}h"),
                ("🚀", "运行状态", status_disp),
                ("🌐", "登入 IP", f"<code>{login_ip}</code>")
            ])
        elif "5 d" in str(renew_res.get('message', '')) or (before_h > 108):
            send_notice("maxed_out", [
                ("🖥️", "服务器名称", current_server_name),
                ("🆔", "ID", f"<code>{server_id}</code>"),
                ("⏰", "剩余时间", f"{after_h}h"),
                ("🚀", "运行状态", status_disp),
                ("💡", "提示", "已近120h上限，暂无需续期。"),
                ("🌐", "登入 IP", f"<code>{login_ip}</code>")
            ])
        else:
            send_notice("renew_failed", [("🖥️", "服务器名称", current_server_name), ("💡", "原因", renew_res.get('message','未知错误'))])

    except Exception as e:
        send_notice("error", [("🖥️", "服务器", current_server_name), ("❌", "故障", f"<code>{str(e)[:100]}</code>")])
    finally:
        if driver: driver.quit()

if __name__ == "__main__":
    run_task()
